### DISCUSSION TITLE
Extract values from array inside json

### DISCUSSION THREAD

**Discussion by koder29406** [schibsted/jslt#337] [Q&A] [answered]
Hello!
I'm trying to convert a JOLT spec to JLST spec. The following worked with JOLT, but I don't get how to get jlst to do the same thing.
Input:
```
{
  "sourceId": "SN98640:0",
  "referenceTime": "2024-04-01T23:00:00.000Z",
  "observations": [
    {
      "elementId": "air_temperature",
      "value": -9.4,
      "exposureCategory": "2"
    },
    {
      "elementId": "wind_speed",
      "value": 8.3,
      "exposureCategory": "2"
    }
  ]
}
```

Desired output, which I get from JOLT:
```
{
  "stasjonsid" : "SN98640:0",
  "tid" : "2024-04-01T23:00:00.000Z",
  "temperatur" : -9.4,
  "vindstyrke" : 8.3
}
```

Working JOLT spec:
```
[
  {
    "operation": "shift",
    "spec": {
      "sourceId": "stasjonsid",
      "referenceTime": "tid",
      "observations": {
        "*": {
          "value": "@(1,elementId)"
        }
      }
    }
  },
  {
    "operation": "shift",
    "spec": {
      "air_temperature": "temperatur",
      "wind_speed": "vindstyrke",
      "referenceTime": "tid",
      "*": {
        "@": "&"
      }
    }
  }
]
```

Any suggestions how to fix this?

**Comment by larsga** [comment]
The way to write a transform like this in JSLT is to start with the output you want. Use that as the starting point. Then go through the values and replace them with a JSLT expression that will find the correct value in the input. Use [the playground](https://www.garshol.priv.no/jslt-demo) to try it out as you go. If you keep the tutorial open you should be able to do solve the first two pretty much immediately.

**Reply by koder29406** [reply]
Thanks, I read the examples earlier and got this far:
```
{
  "stasjonsid" : .sourceId,
  "tid" : .referenceTime,
  "observations" : {for (.observations)
    .elementId : .value
  }
}
```

```
{
  "stasjonsid" : "SN98640:0",
  "tid" : "2024-04-01T23:00:00.000Z",
  "observations" : {
    "air_temperature" : -9.4,
    "wind_speed" : 8.3
  }
}
```
I can't get the temperatue and wind to be on the root level instead of inside observations.


It feels like this spec should work, but I get a parse error.
```
{
  "stasjonsid" : .sourceId,
  "tid" : .referenceTime,
  { for ( .observations) .elementId : .value }
}
```

**Reply by larsga** [reply, marked as answer]
There's a few different ways you can deal with `observations`. The big question is: is the order reliable? It probably isn't.

You need to first find the observation that has "air_temperature", and only afterwards can you get the value. Try this: 

```
[for (.observations) .value if (.elementId == "air_temperature")]
```

It will get you an array. Now you need to pick out the first element from it. Unfortunately, you can't just tack on a `[0]`.

```
let air_temperatures = [for (.observations) .value if (.elementId == "air_temperature")]
$air_temperatures[0]
```

will do it.

Put the two `let`s at the top and you'll have it.

**Comment by koder29406** [comment]
Thanks, this worked!
```
let air_temperatures = [for (.observations) .value if (.elementId == "air_temperature")]
let vind = [for (.observations) .value if (.elementId == "wind_speed")]

{
  "stasjonsid" : .sourceId,
  "tid" : .referenceTime,
  "temperatur" : $air_temperatures[0],
  "vind" : $vind[0]
}
```

What if i had more measurements than temperature and wind. What would be a good generic way to get for example all elementIds and values?

**Reply by larsga** [reply]
You kind of did get all of them already with this:

```
  "observations" : {for (.observations)
    .elementId : .value
  }
```

What you could do is make one object with the first two fields, then one that has all observations, and `+` them together into a single object:

```
{
  "stasjonsid" : .sourceId,
  "tid" : .referenceTime
 } + {
  for (.observations)
    .elementId : .value
}
```

Of course, this gives you the last part in the wrong vocabulary.  You could define a function that translates the `elementId` into the right term. One way to do that is just with `if` statements, but you could also do `let mapping = {}` and then use `get-key()` to get the mapping.

**Comment by koder29406** [comment]
Thank you for your help. I'll take a look at that later!
